### ISSUE TITLE
Detect violations to Custom XML rules 

### ISSUE THREAD

**Issue by rsoesemann** [ChuckJonas/vscode-apex-pmd#96]
The latest versions of PMD now support Custom Rules for XML metadata. Which is a super helpful feature for Salesforce Developers using PMD and PMD related IDE extensions as Salesforce store most of the "code" as XML files.

I am kindly asking to extend this extension to support that. Here is proof that it doesn't work with the current version.

### WORKS
My repo contains a **Custom XPath rule for Apex** that detects illegal use of `System.debug`:
https://github.com/rsoesemann/unhappy-soup/blob/master/ruleset.xml#L26

ApexPMD perfectly detects such an issue in 
https://github.com/rsoesemann/unhappy-soup/blob/master/force-app/main/default/classes/Voldemort.cls#L25

### DOESNT WORK
My repo also contains a **Custom XPath rule for XML** that detect unsafe Security Settings:
https://github.com/rsoesemann/unhappy-soup/blob/master/ruleset.xml#L57

Apex PMD does NOT find the problematic line in 
https://github.com/rsoesemann/unhappy-soup/blob/master/force-app/main/default/settings/Security.settings-meta.xml#L12

**Comment by ChuckJonas** [comment]
@rsoesemann I just upgraded pmd bin from `6.37` to 6.41`.

These are rules to run against XML metadata?  Can you provide a text version of some of these rules for testing?

The extension might need to be updated to activate on XML code...

**Comment by rsoesemann** [comment]
Sure @ChuckJonas lets just use this repo of mine to iterate on such topics

https://github.com/rsoesemann/unhappy-soup

There is metadata with issues and a `ruleset.xml` which should detect them.

**Comment by rsoesemann** [comment]
@ChuckJonas did you see my response or did I miss a point you are waiting for?

**Comment by ChuckJonas** [comment]
@rsoesemann sorry for the delayed response.  While I agree this would be a great addition, it's not something I'm actively working on.

I do not personally use PMD anymore and our salesforce team does not currently use any x-path rules. 

Luckily, this is an open source project and I'd be more than happy to accept a PR for this feature.  

If it's not something you are personally interested in contributing, then I'd encourage you to add reproducible steps to this ticket (ideally a single rule + metadata file with "expected outcome").  That would make it much easier for someone else to pick up (maybe I can find a dev on my team who is interested).

**Comment by rsoesemann** [comment]
@ChuckJonas I understand that you are not activly contributing to it. I don't feel well versed to contribute to a VS plugin repo but I will try to improve the level of detail in the issue and hope that somebody else will eventually pick it.

**Comment by rsoesemann** [comment]
@nawforce just saw you being a contributor here as well. Would you mind helping a bit here? Maybe this is really just enhancing the file extensions as Chuck mentioned.

**Comment by ChuckJonas** [comment]
@rsoesemann thanks, I was able to reproduce with your new detailed steps above.

I updated the `package.json` to enable running the commands against XML.  Unfortunately, when I ran it, I got the following error:

```

PMD Command: /usr/bin/java -cp "/Users/charlesjonas/Documents/code/__vscode/vscode-apex-pmd/bin/pmd/lib/*:/Users/charlesjonas/Documents/code/__examples/my-sandbox/*" net.sourceforge.pmd.PMD -f csv --cache "/Users/charlesjonas/Documents/code/__examples/my-sandbox/.pmdCache" -d "/Users/charlesjonas/Documents/code/__examples/my-sandbox/force-app/main/default/settings/Security.settings-meta.xml" -R "/Users/charlesjonas/Documents/code/__examples/my-sandbox/rules.xml"
stdout:
Mandatory arguments:
1) A java source code filename or directory
2) A report format 
3) A ruleset filename or a comma-delimited string of ruleset filenames

For example: 
C:\>pmd-bin-6.41.0\bin\pmd.bat -d c:\my\source\code -f html -R java-unusedcode

Languages and version suported:
apex, vf

Available report formats and their configuration properties are:
   codeclimate: Code Climate integration.
   csv: Comma-separated values tabular format.
        problem - Include Problem column   default: true
        package - Include Package column   default: true
        file - Include File column   default: true
        priority - Include Priority column   default: true
        line - Include Line column   default: true
        desc - Include Description column   default: true
        ruleSet - Include Rule set column   default: true
        rule - Include Rule column   default: true
   emacs: GNU Emacs integration.
   empty: Empty, nothing.
   html: HTML format
        linePrefix - Prefix for line number anchor in the source file.
        linkPrefix - Path to HTML source.
        htmlExtension - Replace file extension with .html for the links (default: false)   default: false
   ideaj: IntelliJ IDEA integration.
        classAndMethodName - Class and Method name, pass '.method' when processing a directory.   default: 
        sourcePath - Source path.   default: 
        fileName - File name.   default: 
   json: JSON format.
   sarif: Static Analysis Results Interchange Format (SARIF)
   summaryhtml: Summary HTML format.
        linePrefix - Prefix for line number anchor in the source file.
        linkPrefix - Path to HTML source.
        htmlExtension - Replace file extension with .html for the links (default: false)   default: false
   text: Text format.
   textcolor: Text format, with color support (requires ANSI console support, e.g. xterm, rxvt, etc.).
        color - Enables colors with anything other than 'false' or '0'.   default: yes
   textpad: TextPad integration.
   vbhtml: Vladimir Bossicard HTML format.
   xml: XML format.
        encoding - XML encoding format, defaults to UTF-8.   default: UTF-8
   xslt: XML with a XSL Transformation applied.
        encoding - XML encoding format, defaults to UTF-8.   default: UTF-8
        xsltFilename - The XSLT file name.
   yahtml: Yet Another HTML format.
        outputDir - Output directory.

For example on windows: 
C:\>pmd-bin-6.41.0\bin\pmd.bat -dir c:\my\source\code -format text -R rulesets/java/quickstart.xml -version 1.5 -language java -debug
C:\>pmd-bin-6.41.0\bin\pmd.bat -dir c:\my\source\code -f xml -rulesets rulesets/java/quickstart.xml,category/java/codestyle.xml -encoding UTF-8
C:\>pmd-bin-6.41.0\bin\pmd.bat -d c:\my\source\code -rulesets rulesets/java/quickstart.xml -auxclasspath lib\commons-collections.jar;lib\derby.jar
C:\>pmd-bin-6.41.0\bin\pmd.bat -d c:\my\source\code -f html -R rulesets/java/quickstart.xml -auxclasspath file:///C:/my/classpathfile

For example on *nix: 
$ pmd-bin-6.41.0/bin/run.sh pmd -dir /home/workspace/src/main/java/code -f html -rulesets rulesets/java/quickstart.xml,category/java/codestyle.xml
$ pmd-bin-6.41.0/bin/run.sh pmd -d ./src/main/java/code -R rulesets/java/quickstart.xml -f xslt -property xsltFilename=my-own.xsl
$ pmd-bin-6.41.0/bin/run.sh pmd -d ./src/main/java/code -f html -R rulesets/java/quickstart.xml -auxclasspath commons-collections.jar:derby.jar

stderr:Unknown Language 'xml' for rulea0H5g000002bou5EAA, supported Languages are apex, vf
```

Any idea?

> stderr:Unknown Language 'xml' for rulea0H5g000002bou5EAA, supported Languages are apex, vf 

seems like the culprit.

**Comment by rsoesemann** [comment]
@ChuckJonas Did you try if the same setting works when you run PMD locally?

**Comment by rsoesemann** [comment]
@adangel as I have seen you respond to a similar issue years ago maybe you can see that both are related and can be solved similarly.

As Gradle PMD Plugin in 2018 https://github.com/gradle/gradle/issues/5451 this Salesforce Specific VSCode PMD extension "fails" to use and detect the new Custom Rules on Salesforce XML-based metadata.

**Comment by adangel** [comment]
Yes, you need additionally

```xml
    <dependency>
      <groupId>net.sourceforge.pmd</groupId>
      <artifactId>pmd-xml</artifactId>
      <version>${pmd.version}</version>
    </dependency>
```

**Comment by nawforce** [comment]
Created a PR for this, pretty straight forward given the earlier investigation. A couple of things worth noting:
1. If I used an invalid XPath I did not get an error, I was expecting PMD to report something.
2. The Salesforce XML parser is known to allow some nonconforming documents e.g. allowing the use of & rather than &amp;. In my experience these are fairly rare but we wondering if that needs documenting somewhere.
